sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala

```diff
@@ -351,11 +351,12 @@ class ParquetFileFormat
       if (pushed.isDefined) {
         ParquetInputFormat.setFilterPredicate(hadoopAttemptContext.getConfiguration, pushed.get)
       }
+      val taskContext = Option(TaskContext.get())
       val parquetReader = if (enableVectorizedReader) {
         val vectorizedReader = new VectorizedParquetRecordReader()
         vectorizedReader.initialize(split, hadoopAttemptContext)
         logDebug(s"Appending $partitionSchema ${file.partitionValues}")
-        vectorizedReader.initBatch(partitionSchema, file.partitionValues)
+        vectorizedReader.initBatch(partitionSchema, file.partitionValues, taskContext.isDefined)
```

Comment:
When `taskContext.isDefined` is false, we force to use on-heap.
This is because we cannot register `iter.close` at [here](https://github.com/apache/spark/pull/17436/files#diff-ee26d4c4be21e92e92a02e9f16dbc285R380) to release off-heap.